### DISCUSSION TITLE
Debugging

### DISCUSSION THREAD

**Discussion by CharlesCZ** [sventorben/keycloak-home-idp-discovery#187] [Q&A] [answered]
Hi, how do you debug it? I would like to debug this Keycloak authenticator but when I run annotated @Testcontainers tests. Then it doesn't stop on the code I want to investigate. Do you have instruction?

**Comment by sventorben** [comment]
Hello @CharlesCZ,

before I answer your question, would you mind letting me know why you would like to debug the extension? Do you experience a bug or anything like that?

I generally use one of two options to debug the extension:

1) Use the docker-compose file in the root directory and attach a remote debugger. You can build the extension with `mvn clean install` and start a docker container via `docker-compose up`. You then get a running Keycloak instance on localhost port 8080. The debug port will be available on localhost and port 8787. 
2) If I need to debug a concrete test, I add the debugging options (see docker-compose.yml) to the config options in `KeycloakDockerContainer` class. Then I start the test and attach a remote debugger as in option 1.

**Reply by CharlesCZ** [reply]
How exactly do you override settings of KeycloakDockerContainer so that It doesn't create new container but reuses this from docker-compose?

**Reply by sventorben** [reply, marked as answer]
The way the project is set up, it does not support this out of the box. You would need to add a custom network in docker compose file. You would need to change the network definition in tests to match the one in docker compose file.  Then you shoul disable usage of KeycloakDockerContainer, because you already have a running Container via docker compose. Make sure you set the correct URL in AuthenticatorConfig and for the Keycloak admin client.

**Comment by CharlesCZ** [comment]
Hi, thanks for response I need a feature which provides username to redirected login screen. I don't want to duplicate username typing. So when I write username in keycloak login screen I want this info to be retained in e.g. github login screen. It would be best to have it provided to all available idp. I noticed however that in the case of github to provide username I need to append login= request parameter (login_hint doesn't work there from what I can see). So It may differ in the case of different idps. I thought that maybe in this plugin I could provide it.

**Reply by sventorben** [reply]
This feature is already implemented in Keycloak. You need to enable forwarding of the `login_hint` parameter in identity provider configs. The config option is called `Pass login_hint` for OIDC or `Pass subject` for SAML providers.

If the Github provider needs a parameter different from `login_hint`, this logic should be implented in the Keycloak project and not in this extension. 

The code for the github identity provider is located at [GitHubIdentityProvider](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/social/github/GitHubIdentityProvider.java). It extends the [AbstractOauth2IdentityProvider](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/broker/oidc/AbstractOAuth2IdentityProvider.java#L329) that sets the `login_hint` parameter. 

I suggest you file an issue at the Keycloak project and contribute a solution there. Make sure you check the [contributions guide](https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md).

**Reply by CharlesCZ** [reply]
Thank you for your response. Do you know whether is it possible to set 'Pass login_hint' via UI? I cannot find it anywhere in Identity providers section

**Reply by CharlesCZ** [reply]
![Screenshot from 2023-04-27 16-50-02](https://user-images.githubusercontent.com/37272177/234900508-3753d21f-a73d-463f-8e28-c6135a92789c.png)

**Reply by sventorben** [reply]
What kind of Identity Provider is this? The Github provider?

I am not sure if it was there in the old admin console. It may be an issue with the new admin console. See for example this ticket here which is asking for similiar functionality: https://github.com/keycloak/keycloak/issues/15344

I think the filed you are looking for is defined at https://github.com/keycloak/keycloak/blob/21.1.1/js/apps/admin-ui/src/identity-providers/add/ExtendedNonDiscoverySettings.tsx#L41, but it's not shown because it does not qualify as an oidc provider.
See https://github.com/keycloak/keycloak/blob/21.1.1/js/apps/admin-ui/src/identity-providers/add/DetailSettings.tsx#L344 and https://github.com/keycloak/keycloak/blob/21.1.1/js/apps/admin-ui/src/identity-providers/add/DetailSettings.tsx#L286

**Reply by CharlesCZ** [reply]
Yup, it's Github provider
